### PR TITLE
Fix ortmodule for the pytorch model with ort device

### PR DESCRIPTION
**Pull request by souptc**
**Description**: 
two issues there:
1. the ORTModule didn't register the EP used in ort device
2. The tensor conversion from ort aten tensor to ort value doesn't need dlpack because we can decode it from ort eager mode.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py [hunk 1]
```diff
@@ -177,8 +176,7 @@ def backward(ctx, *grad_outputs):
                 initializer_index = num_user_input_grads
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_training_manager.py:176]: Is initializer_index` dead code after this change? Is there any performance penalty with that change? I remember some manuveurs surrounding vectors were done for performance reasons

File: tools/ci_build/github/windows/eager/requirements.txt [hunk 2]
```diff
@@ -3,3 +3,8 @@ wheel
 numpy
 typing_extensions
 torch
+cerberus
```
Comment (thiagocrepaldi) [inline tools/ci_build/github/windows/eager/requirements.txt:6]: `cerberus` is used for training, only. Specifically for ORTTrainer. Are we supportng this in this pipeline?

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py (hunk 1 above)
Comment (souptc) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py:176]: why it is dead code? i don't see why my change related with initializer_index.

File: tools/ci_build/github/windows/eager/requirements.txt (hunk 2 above)
Comment (souptc) [reply tools/ci_build/github/windows/eager/requirements.txt:6]: yes this is an accident, will remove it

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py (hunk 1 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py:176]: My bad, didn't notice line 180 still uses it

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 3]
```diff
@@ -348,7 +348,7 @@ std::unordered_map<std::string, std::unordered_map<std::string, py::object>> Con
 void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn ep_registration_fn) {
   py::class_<std::vector<OrtValue>>(m, "OrtValueVector")
       .def(py::init<>())
-      .def("push_back", [](std::vector<OrtValue>* v, const OrtValue& ortvalue) {
+      .def("push_back", [](std::vector<OrtValue>* v, const OrtValue& ortvalue, const bool /*to_make_api_consistant*/) {
```
Comment (xadupre) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:351]: Consistent with which API?

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py [hunk 4]
```diff
@@ -163,8 +163,7 @@ def backward(ctx, *grad_outputs):
                 for input_name in self._graph_info.user_input_names:
                     # Append to the results the backward output for each input that required grad
                     if input_name in require_grad_names_set:
-                        results.append(_utils._torch_tensor_from_dl_pack(
-                            backward_outputs.dlpack_at(require_grad_names_index),
+                        results.append(_utils._ortvalue_to_torch_tensor(
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_training_manager.py:166]: Is this OK for non-eager mode? Is `backward_outputs[require_grad_names_index]` of type `ORTValue` for non-eager mode?

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py [hunk 5]
```diff
@@ -177,8 +176,7 @@ def backward(ctx, *grad_outputs):
                 initializer_index = num_user_input_grads
                 for initializer_name in self._graph_info.initializer_names:
                     if initializer_name in self._graph_initializer_names_to_train:
-                        results.append(_utils._torch_tensor_from_dl_pack(
-                            backward_outputs.dlpack_at(initializer_index),
+                        results.append(_utils._ortvalue_to_torch_tensor(
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_training_manager.py:179]: Same as above

File: orttraining/orttraining/python/training/ortmodule/_utils.py [hunk 6]
```diff
@@ -38,21 +38,24 @@ def _ortvalue_from_torch_tensor(torch_tensor):
...
     # PyTorch's to_dlpack() uses same config for both torch.bool and torch.uint8,
     # and convert the config to torch.uint8 tensor duing from_dlpack().
     # So we need to convert the torch tensor to torch.bool type if OrtValue is bool tensor.
     dlpack_tensor = ortvalue.to_dlpack()
-    return _torch_tensor_from_dl_pack(dlpack_tensor, ortvalue, device)
+    return _torch_tensor_from_dl_pack(dlpack_tensor, ortvalue)
 
-def _torch_tensor_to_dlpack(tensor):
+# convert torch tensor to ort accept tensors, could be dlpack or ortvalue
+def _torch_tensor_to_ort_accept_tensors(tensor):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_utils.py:56]: Having a function that either returns a `ORTValue` or a `DLPack` tensor is confusing.
Ideally it should return the same type in both branches, so callers know exactly what to expect

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 3 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:351]: Is this really needed? I don't think I saw the third argument being passed in in any call in this PR

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py (hunk 4 above)
Comment (souptc) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py:166]: the result is not with type ORTValue, it is with type aten::tensor, for both eager mode and non-eager mode.
The previously implementation is convert the output to dlpack at line 167 then invoke "_torch_tensor_from_dlpack" to convert to aten::tensor. For eager mode we don't need to use dlpack as the bridge. the change here is just consolidate them under the same function, there is no input/output type change.

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py (hunk 5 above)
Comment (souptc) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py:179]: same as the explanation above.

File: orttraining/orttraining/python/training/ortmodule/_utils.py (hunk 6 above)
Comment (souptc) [reply orttraining/orttraining/python/training/ortmodule/_utils.py:56]: I can move the branches to the caller to eliminate the confusing, but that will make the caller looks more complicated.

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 3 above)
Comment (souptc) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:351]: it is used here:
https://github.com/microsoft/onnxruntime/blob/e2ee7a734e7b4acd95ef6d27e753411afcdb8657/orttraining/orttraining/python/training/ortmodule/_training_manager.py#L51

but as discussed in other places, if we don't want the "to_ort_accepted_tensors" return different types, and move the branches to the caller, then we don't need this,

File: orttraining/orttraining/python/training/ortmodule/_utils.py (hunk 6 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule/_utils.py]: Overall changes are OK, my only concern is the multi-return aspect of `_torch_tensor_to_ort_accept_tensors`

Having conditional return types probably kill JIT's compiler optimization and in general as a bad idea for the reasons many languages use static typing, such as C++. Having to refactor an API to fit this duality really increases coupling and makes maintenance hard down the line.

I'd consider simplifying implementation to use specific functions for specific tasks and use external logic/pattern to remove ambiguity

Comment (souptc) [reply orttraining/orttraining/python/training/ortmodule/_utils.py]: update the implementation, to make it a explicit condition in the caller.

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 3 above)
Comment (souptc) [reply orttraining/orttraining/python/orttraining_pybind_state.cc]: removed.

File: orttraining/orttraining/python/training/ortmodule/_utils.py [hunk 7]
```diff
@@ -38,33 +38,33 @@ def _ortvalue_from_torch_tensor(torch_tensor):
...
 
-def _torch_tensor_from_dl_pack(dlpack, ortvalue, device):
-    torch_tensor = from_dlpack(dlpack) if device.type != 'ort' else C.ort_from_dlpack(dlpack)
+def _torch_tensor_from_dl_pack(dlpack, ortvalue):
+    torch_tensor = from_dlpack(dlpack)
     return torch_tensor.to(torch.bool) if ortvalue.data_type() == 'tensor(bool)' else torch_tensor
 
 
 def _ortvalue_to_torch_tensor(ortvalue, device):
+    if device.type == 'ort':
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_utils.py:47]: Is `at:Tensor` the same as `torch.Tensor`? If not, we have the same multi-type return issue here

Comment (souptc) [reply orttraining/orttraining/python/training/ortmodule/_utils.py:47]: the same

**Review by thiagocrepaldi — APPROVED** [review]
LGTM

**Pull request merged**